M2/Macaulay2/m2/methods.m2

```diff
@@ -516,38 +505,89 @@ dispatcherFunctions = join (dispatcherFunctions, {
 
 -----------------------------------------------------------------------------
 -- hooks
-
-addHook = method()
+-- also see hooks in code.m2
+-- TODO: get this to also work with methods like ((_, =), Y, Z) or ((SPACE, =), Function, Thing)
+
+protect symbol Hooks
+protect symbol HookAlgorithms
+protect symbol HookPriority
+
+getHookStore = (key, create) -> (
+    -- retrieve (or create) the mutable hashtable of Hooks installed on a symbol or hashtable
+    sym := if key#?0 then key#0 else error "createHookStore: encountered empty method key";
+    sym  = if instance(sym, Symbol) then sym else getGlobalSymbol toString sym; -- TODO: is toString the right function?
+    -- this branch depends on whether hooks are installed on a symbol or a hash table
+    if #key == 1 then ( key = sym; if value sym =!= sym then value sym else if create then sym <- new MutableHashTable from {} ) else (
+	obj := youngest drop(key, 1);
+	if instance(obj, MutableHashTable) then ( if       obj.?Hooks then       obj.Hooks else if create then       obj.Hooks = new MutableHashTable from {} ) else
+	if instance(obj, HashTable)        then ( if obj.cache.?Hooks then obj.cache.Hooks else if create then obj.cache.Hooks = new MutableHashTable from {} )))
```

Comment:
When can this happen?

Reply:
Ahah.  It can't occur because `instance(obj, HashTable)` is always false.  That's because `youngest` returns a mutable hash table.  Also, `instance(obj, MutableHashTable)` is always true.  So you can simplify those two lines considerably -- first test for `obj =!= null`.